libraries/shared/storage/fetcher/geth_rpc_storage_fetcher.go

```diff
@@ -67,45 +64,31 @@ func (fetcher GethRpcStorageFetcher) FetchStorageDiffs(out chan<- types.RawDiff,
 		case err := <-clientSubscription.Err():
 			logrus.Errorf("error with client subscription: %s", err.Error())
 			errs <- err
-		case diff := <-ethStatediffPayloadChan:
-			logrus.Trace("received a statediff")
-			stateDiff := new(statediff.StateDiff)
-			decodeErr := rlp.DecodeBytes(diff.StateDiffRlp, stateDiff)
-			logrus.Tracef("received a statediff from block: %v", stateDiff.BlockNumber)
-			if decodeErr != nil {
-				logrus.Warn("Error decoding state diff into RLP: ", decodeErr)
-				errs <- decodeErr
-			}
-
-			accounts := getAccountsFromDiff(*stateDiff)
-			logrus.Trace(fmt.Sprintf("iterating through %d accounts on stateDiff for block %d", len(accounts), stateDiff.BlockNumber))
-			for _, account := range accounts {
-				logrus.Trace(fmt.Sprintf("iterating through %d Storage values on account", len(account.Storage)))
-				for _, accountStorage := range account.Storage {
-					diff, formatErr := fetcher.formatDiff(account, stateDiff, accountStorage)
-					logrus.Tracef("adding storage diff to out channel. keccak of address: %v, block height: %v, storage key: %v, storage value: %v",
-						diff.HashedAddress.Hex(), diff.BlockHeight, diff.StorageKey.Hex(), diff.StorageValue.Hex())
-					if formatErr != nil {
-						errs <- formatErr
-					}
-
-					out <- diff
-				}
-			}
+		case diffPayload := <-ethStatediffPayloadChan:
+			logrus.Trace("received a statediff payload")
+			fetcher.handleDiffPayload(diffPayload, out, errs)
 		}
-
 	}
 }
 
-func (fetcher GethRpcStorageFetcher) formatDiff(account statediff.AccountDiff, stateDiff *statediff.StateDiff, storage statediff.StorageDiff) (types.RawDiff, error) {
-	if fetcher.gethVersion == OldGethPatch {
-		return types.FromOldGethStateDiff(account, stateDiff, storage)
-	} else {
-		return types.FromNewGethStateDiff(account, stateDiff, storage)
+func (fetcher GethRpcStorageFetcher) handleDiffPayload(payload filters.Payload, out chan<- types.RawDiff, errs chan<- error) {
+	var stateDiff filters.StateDiff
+	decodeErr := rlp.DecodeBytes(payload.StateDiffRlp, &stateDiff)
+	if decodeErr != nil {
+		errs <- fmt.Errorf("error decoding storage diff from geth payload: %w", decodeErr)
+		return
 	}
-}
 
-func getAccountsFromDiff(stateDiff statediff.StateDiff) []statediff.AccountDiff {
-	accounts := append(stateDiff.CreatedAccounts, stateDiff.UpdatedAccounts...)
-	return append(accounts, stateDiff.DeletedAccounts...)
+	for _, account := range stateDiff.UpdatedAccounts {
+		logrus.Infof(processingDiffsLogString, len(account.Storage), common.Bytes2Hex(account.Key))
+		for _, accountStorage := range account.Storage {
+			rawDiff, formatErr := types.FromGethStateDiff(account, &stateDiff, accountStorage)
+			if formatErr != nil {
+				errs <- formatErr
```

Comment:
do we potentially want to return after this? Not sure if sending an error will lead the caller to try and close the channel (leading to a panic on subsequent sends)

Reply:
This piece is a little fuzzy for me - I think you're right, that if we were to send an error to the `errs` channel, then it's possible that the next accountStorage we're iterating through could try to send a diff (or an error) to a channel that's already been closed... I think. Also, it probably makes sense to stop processing diffs if the system is going to shut down anyway.